plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoConnectorTest.java

```diff
@@ -406,24 +405,215 @@ public void testHighPrecisionDecimalPredicate()
+            assertPredicatePushdown(table.getName(), "id = 1 OR discount <= NULL");
+            assertPredicatePushdown(table.getName(), "id = 1 OR discount >= NULL");
+
+            assertPredicatePushdown(table.getName(), "id = 1 OR NULL = discount");
+            assertPredicatePushdown(table.getName(), "id = 1 OR NULL != discount");
+            assertPredicatePushdown(table.getName(), "id = 1 OR NULL < discount");
+            assertPredicatePushdown(table.getName(), "id = 1 OR NULL > discount");
+            assertPredicatePushdown(table.getName(), "id = 1 OR NULL <= discount");
+            assertPredicatePushdown(table.getName(), "id = 1 OR NULL >= discount");
+
+            assertPredicatePushdown(table.getName(), "id = 1 OR discount IS NULL");
+            assertPredicatePushdown(table.getName(), "id = 1 OR discount IS NOT NULL");
+
+            assertPredicatePushdown(table.getName(), "id = 1 OR discount IN (NULL, 20)");
+            assertPredicatePushdown(table.getName(), "id = 1 OR 20 IN (NULL, discount)");
+            assertPredicatePushdown(table.getName(), "id = 1 OR NULL IN (discount, 20)");
+
+            assertPredicatePushdown(table.getName(), "id = 1 OR discount NOT IN (NULL, 20)");
+            assertPredicatePushdown(table.getName(), "id = 1 OR 20 NOT IN (NULL, discount)");
+            assertPredicatePushdown(table.getName(), "id = 1 OR NULL NOT IN (discount, 20)");
+        }
+    }
+
+    @Test
+    public void testComplexExpressionPredicatePushdownWithMismatchedDataType()
+    {
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_complex_expression_predicate_pushdown_with_mismatched_data_type",
+                "(id INT, discount INT)")) {
+            Document document1 = new Document("id", 1).append("discount", 8);
+            Document document2 = new Document("id", 2).append("discount", "9");
+            Document document3 = new Document("id", 3).append("discount", 11);
+            Document document4 = new Document("id", 4).append("discount", "16");
+
+            MongoCollection<Document> collection = client.getDatabase(getSession().getSchema().orElseThrow()).getCollection(table.getName());
+            collection.insertMany(ImmutableList.of(document1, document2, document3, document4));
+
+            assertPredicatePushdown(table.getName(), "id = 1 OR discount = 11");
+            assertThat(query("SELECT id FROM %s WHERE id = 1 OR discount != 10".formatted(table.getName())))
+                    .matches("VALUES 1, 2, 3, 4"); // TODO Enabling complex filter pushdown returns the mismatched datatype document as well
+            assertPredicatePushdown(table.getName(), "id = 1 OR discount < 11");
+            assertThat(query("SELECT id FROM %s WHERE id = 1 OR discount > 10".formatted(table.getName())))
+                    .matches("VALUES 1, 2, 3, 4"); // TODO Enabling complex filter pushdown returns the mismatched datatype document as well
```

Comment:
I hesitate to approve the change having correctness issues even if it's disabled by default. Is it difficult to avoid it by narrowing the scope of this PR more?

Reply:
Thank for the review @ebyhr, Currently, I am working on this to fix all correctness issue in this PR.

Reply:
I fixed all these correctness issues by using "$type" expression.